### PR TITLE
Only use immutable tls secret

### PR DESCRIPTION
**Pull request by Kostov6**
**How to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:
This PR forces referenced tls secret resource to be immutable by rejecting shoot config if mutable

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener-extension-shoot-rsyslog-relp/issues/2

**Special notes for your reviewer**:

**Release note**:

```breaking user
When changing referenced TLS secret in `shoot.spec.resources[]` the user should provide only immutable secret
```

### CODE REVIEW

**Review by plkokanov — CHANGES_REQUESTED** [review]
Some comments inline.
Additionally, you should flag this as a breaking change.

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ func ValidateRsyslogRelpSecret(secret *corev1.Secret) error {
 	if _, ok := secret.Data[constants.RsyslogPrivateKeyKey]; !ok {
 		return fmt.Errorf("secret %s is missing %s value", key.String(), constants.RsyslogPrivateKeyKey)
 	}
+	if secret.Immutable == nil || !*secret.Immutable {
```
[the remark this answers is not in the document]
Comment (Kostov6) [reply pkg/utils/utils.go:37]: Shouldn't it be `if !ptr.Deref(secret.Immutable, false) {`?

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -34,6 +35,9 @@ func ValidateRsyslogRelpSecret(secret *corev1.Secret) error {
 	if _, ok := secret.Data[constants.RsyslogPrivateKeyKey]; !ok {
 		return fmt.Errorf("secret %s is missing %s value", key.String(), constants.RsyslogPrivateKeyKey)
 	}
+	if !ptr.Deref(secret.Immutable, false) {
```
[the remark this answers is not in the document]
Comment (Kostov6) [reply pkg/utils/utils.go:38]: Should the example secret in the document have a hash suffix?

[the remark this answers is not in the document]
Comment (ialidzhikov) [reply pkg/utils/utils.go:38]: It is already suffixed with v1 so I assume this v1 represents the revision of the Secret. At least this is also the naming pattern we use in the registry-cache docs when dealing with immutable Secrets: see the docs in https://github.com/gardener/gardener-extension-registry-cache/blob/9176c6572d582e0725024b1db30735d8c7f25cbc/docs/usage/registry-cache/upstream-credentials.md, you will find there `-v1` and `-v2` suffixes in the Secret names.

File: docs/usage/configuration.md [hunk 3]
```diff
@@ -123,16 +123,19 @@ loggingRules:
 
 ### Securing the Communication to the Target Server with TLS
 
-The communication to the target server is not encrypted by default. To enable encryption, set the `.tls.enabled` field in the `shoot-rsyslog-relp` extension configuration to `true`. In this case, a secret which contains the TLS certificates used to establish the TLS connection to the server must be created in the same project namespace as your Shoot.
+The communication to the target server is not encrypted by default. To enable encryption, set the `.tls.enabled` field in the `shoot-rsyslog-relp` extension configuration to `true`. In this case, a immutable secret which contains the TLS certificates used to establish the TLS connection to the server must be created in the same project namespace as your Shoot.
```
Comment (plkokanov) [inline docs/usage/configuration.md:126]: nit:
```suggestion
The communication to the target server is not encrypted by default. To enable encryption, set the `.tls.enabled` field in the `shoot-rsyslog-relp` extension configuration to `true`. In this case, an immutable secret which contains the TLS certificates used to establish the TLS connection to the server must be created in the same project namespace as your Shoot.
```

File: docs/usage/configuration.md [hunk 4]
```diff
@@ -123,16 +123,19 @@ loggingRules:
...
 
 An example Secret is given below:
 
+> **Note:**  The secret should be immutable
```
Comment (plkokanov) [inline docs/usage/configuration.md:130]:
```suggestion
> **Note:**  The secret must be immutable
```

File: pkg/utils/utils.go [hunk 5]
```diff
@@ -34,6 +35,9 @@ func ValidateRsyslogRelpSecret(secret *corev1.Secret) error {
...
+		return fmt.Errorf("secret %s should be immutable", key.String())
```
Comment (plkokanov) [inline pkg/utils/utils.go:39]:
```suggestion
		return fmt.Errorf("secret %s must be immutable", key.String())
```

File: pkg/utils/utils_test.go [hunk 6]
```diff
@@ -43,35 +43,41 @@ var _ = Describe("Utils", func() {
...
 		Entry(
 			"should return error if secret does not contain 'tls' data entry",
-			[]byte("caData"), []byte("crtData"), []byte("tlsData"), []byte("extraData"),
+			[]byte("caData"), []byte("crtData"), []byte("tlsData"), []byte("extraData"), true,
 			MatchError(ContainSubstring("secret foo/rsyslog-secret should have only three data entries")),
 		),
+		Entry(
+			"should return error if secret is mutable",
+			[]byte("caData"), []byte("crtData"), []byte("tlsData"), []byte("extraData"), false,
+			MatchError(ContainSubstring("secret foo/rsyslog-secret should be immutable")),
```
Comment (plkokanov) [inline pkg/utils/utils_test.go:80]:
```suggestion
			MatchError(ContainSubstring("secret foo/rsyslog-secret must be immutable")),
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
Few more minor nits related to grammar and proper terms. 
We could also include this PR in the new release.

**Pull request merged**